File: petclinic-java-gradle-thymeleaf-tailwind-webpack-htmx-ts/src/main/java/dev/pollito/petclinic_java_gradle_thymeleaf_tailwind_webpack_htmx_ts/vet/VetController.java
package dev.pollito.petclinic_java_gradle_thymeleaf_tailwind_webpack_htmx_ts.vet;

import dev.pollito.petclinic_java_gradle_thymeleaf_tailwind_webpack_htmx_ts.specialty.SpecialtyService;
import dev.pollito.petclinic_java_gradle_thymeleaf_tailwind_webpack_htmx_ts.util.WebUtils;
import jakarta.validation.Valid;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.web.PageableDefault;
import org.springframework.data.web.SortDefault;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.servlet.mvc.support.RedirectAttributes;

@Controller
@RequestMapping("/vets")
public class VetController {

    private final VetService vetService;
    private final SpecialtyService specialtyService;

    public VetController(final VetService vetService, final SpecialtyService specialtyService) {
        this.vetService = vetService;
        this.specialtyService = specialtyService;
    }

    @ModelAttribute
    public void prepareContext(final Model model) {
        model.addAttribute("vetSpecialtySpecialtiesValues", specialtyService.getSpecialtyValues());
    }

    @GetMapping
    public String list(@RequestParam(name = "filter", required = false) final String filter,
            @SortDefault(sort = "id") @PageableDefault(size = 20) final Pageable pageable,
            final Model model) {
        final Page<VetDTO> vets = vetService.findAll(filter, pageable);
        model.addAttribute("vets", vets);
        model.addAttribute("filter", filter);
        model.addAttribute("paginationModel", WebUtils.getPaginationModel(vets));
        return "vet/list";
    }

    @GetMapping("/add")
    public String add(@ModelAttribute("vet") final VetDTO vetDTO) {
        return "vet/add";
    }

    @PostMapping("/add")
    public String add(@ModelAttribute("vet") @Valid final VetDTO vetDTO,
            final BindingResult bindingResult, final RedirectAttributes redirectAttributes) {
        if (bindingResult.hasErrors()) {
            return "vet/add";
        }
        vetService.create(vetDTO);
        redirectAttributes.addFlashAttribute(WebUtils.MSG_SUCCESS, WebUtils.getMessage("vet.create.success"));
        return "redirect:/vets";
    }

    @GetMapping("/edit/{id}")
    public String edit(@PathVariable(name = "id") final Integer id, final Model model) {
        model.addAttribute("vet", vetService.get(id));
        return "vet/edit";
    }

    @PostMapping("/edit/{id}")
    public String edit(@PathVariable(name = "id") final Integer id,
            @ModelAttribute("vet") @Valid final VetDTO vetDTO, final BindingResult bindingResult,
            final RedirectAttributes redirectAttributes) {
        if (bindingResult.hasErrors()) {
            return "vet/edit";
        }
        vetService.update(id, vetDTO);
        redirectAttributes.addFlashAttribute(WebUtils.MSG_SUCCESS, WebUtils.getMessage("vet.update.success"));
        return "redirect:/vets";
    }

    @PostMapping("/delete/{id}")
    public String delete(@PathVariable(name = "id") final Integer id,
            final RedirectAttributes redirectAttributes) {
        vetService.delete(id);
        redirectAttributes.addFlashAttribute(WebUtils.MSG_INFO, WebUtils.getMessage("vet.delete.success"));
        return "redirect:/vets";
    }

}

File: petclinic-kt-gradle-thymeleaf-tailwind-webpack-htmx-ts/src/main/kotlin/dev/pollito/petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts/vet/VetController.kt
package dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.vet

import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.specialty.SpecialtyService
import dev.pollito.petclinic_kt_gradle_thymeleaf_tailwind_webpack_htmx_ts.util.WebUtils
import jakarta.validation.Valid
import org.springframework.data.domain.Pageable
import org.springframework.data.web.PageableDefault
import org.springframework.data.web.SortDefault
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.validation.BindingResult
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.ModelAttribute
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.servlet.mvc.support.RedirectAttributes

@Controller
@RequestMapping("/vets")
class VetController(
    private val vetService: VetService,
    private val specialtyService: SpecialtyService,
    private val vetMapper: VetMapper,
) {

    @ModelAttribute
    fun prepareContext(model: Model) {
        model.addAttribute("vetSpecialtySpecialtiesValues", specialtyService.getSpecialtyValues())
    }

    @GetMapping
    fun list(
        @RequestParam(name = "filter", required = false) filter: String?,
        @SortDefault(sort = ["id"]) @PageableDefault(size = 20) pageable: Pageable,
        model: Model,
    ): String {
        val vets = vetService.findAll(filter, pageable)
        model.addAttribute("vets", vets.map { vetMapper.map(it) })
        model.addAttribute("filter", filter)
        model.addAttribute("paginationModel", WebUtils.getPaginationModel(vets))
        return "vet/list"
    }

    @GetMapping("/add")
    fun add(@ModelAttribute("vet") vetDTO: VetDTO): String = "vet/add"

    @PostMapping("/add")
    fun add(
        @ModelAttribute("vet") @Valid vetDTO: VetDTO,
        bindingResult: BindingResult,
        redirectAttributes: RedirectAttributes,
    ): String {
        if (bindingResult.hasErrors()) {
            return "vet/add"
        }
        vetService.create(vetMapper.map(vetDTO))
        redirectAttributes.addFlashAttribute(
            WebUtils.MSG_SUCCESS,
            WebUtils.getMessage("vet.create.success"),
        )
        return "redirect:/vets"
    }

    @GetMapping("/edit/{id}")
    fun edit(@PathVariable(name = "id") id: Int, model: Model): String {
        model.addAttribute("vet", vetMapper.map(vetService.get(id)))
        return "vet/edit"
    }

    @PostMapping("/edit/{id}")
    fun edit(
        @PathVariable(name = "id") id: Int,
        @ModelAttribute("vet") @Valid vetDTO: VetDTO,
        bindingResult: BindingResult,
        redirectAttributes: RedirectAttributes,
    ): String {
        if (bindingResult.hasErrors()) {
            return "vet/edit"
        }
        vetService.update(id, vetMapper.map(vetDTO))
        redirectAttributes.addFlashAttribute(
            WebUtils.MSG_SUCCESS,
            WebUtils.getMessage("vet.update.success"),
        )
        return "redirect:/vets"
    }

    @PostMapping("/delete/{id}")
    fun delete(@PathVariable(name = "id") id: Int, redirectAttributes: RedirectAttributes): String {
        vetService.delete(id)
        redirectAttributes.addFlashAttribute(
            WebUtils.MSG_INFO,
            WebUtils.getMessage("vet.delete.success"),
        )
        return "redirect:/vets"
    }
}
